h2o-algos/src/main/java/hex/rulefit/RuleFit.java

```diff
@@ -307,14 +307,53 @@ SharedTreeModel trainTreeModel(RuleFitModel.Algorithm algorithm, int maxDepth) {
            if (deviance.length < 5) {
                bestLambdaIndex = deviance.length - 1;
            } else {
-               bestLambdaIndex = getBestLambdaIndex(deviance, lambdas);
+               bestLambdaIndex = getBestLambdaIndex(deviance);
+               if (bestLambdaIndex >= lambdas.length) {
+                   bestLambdaIndex = getBestLambdaIndexCornerCase(deviance, lambdas);
+               }
            }
 
             lambdaModel.remove();
             return new double[]{lambdas[bestLambdaIndex]};
         }
 
-        int getBestLambdaIndex(double[] deviance, double[] lambdas) {
+        int getBestLambdaIndex(double[] deviance) {
+            int bestLambdaIndex = deviance.length - 1;
+            if (deviance.length >= 5) {
+                double[] array = differenceArray(signArray(differenceArray(differenceArray(deviance))));
+                for (int i = 0; i < array.length; i++) {
+                    if (array[i] != 0 && i > 0) {
+                        bestLambdaIndex = 3 * i;
+                        break;
+                    }
+                }
+            }
+            return bestLambdaIndex;
+        }
+
+
+        double[] differenceArray(double[] array) {
+            double[] differenceArray = new double[array.length - 1];
+            for (int i = 0; i < array.length - 1; i++) {
+                differenceArray[i] = array[i+1] - array[i];
+            }
+            return differenceArray;
+        }
+
+        double[] signArray(double[] array) {
```

Comment:
```java
        double[] signArray(double[] array) {
            double[] signArray = new double[array.length];
            for (int i = 0; i < array.length; i++) {
                array[i] = Math.signum(array[i]);
            }
            return signArray;
        }
```

Use signum. :wink:  "Každý matematik musí mít svůj herbář funkcí ! Vy nemáte herbář funkcí ?" - doc. Pražák

This piece of code could make it into `ArrayUtils`, WDYT ? :wink: Also, the function name could also be `signum`.

Reply:
:D moved 3522918